src/media-segment-request.js

```diff
@@ -0,0 +1,403 @@
+ *
+ * @param {Object} segment - a simplified copy of the segmentInfo object from SegmentLoader
+ * @param {Function} finishProcessingFn - a callback to execute to continue processing
+ *                                        this request
+ */
+const handleKeyResponse = (segment, finishProcessingFn) => (error, request) => {
+  const response = request.response;
+  const errorObj = handleErrors(error, request);
+
+  if (errorObj) {
+    return finishProcessingFn(errorObj, segment);
+  }
+
+  if (response.byteLength !== 16) {
+    return finishProcessingFn({
+      status: request.status,
+      message: 'Invalid HLS key at URL: ' + request.uri,
+      code: REQUEST_ERRORS.FAILURE,
+      xhr: request
+    }, segment);
+  }
+
+  const view = new DataView(response);
+
+  segment.key.bytes = new Uint32Array([
+    view.getUint32(0),
+    view.getUint32(4),
+    view.getUint32(8),
+    view.getUint32(12)
+  ]);
+  return finishProcessingFn(null, segment);
+};
+
+/**
+ * Handle init-segment responses
+ *
+ * @param {Object} segment - a simplified copy of the segmentInfo object from SegmentLoader
+ * @param {Function} finishProcessingFn - a callback to execute to continue processing
+ *                                        this request
+ */
+const handleInitSegmentResponse = (segment, finishProcessingFn) => (error, request) => {
+  const errorObj = handleErrors(error, request);
+
+  if (errorObj) {
+    return finishProcessingFn(errorObj, segment);
+  }
+  segment.map.bytes = new Uint8Array(request.response);
+  return finishProcessingFn(null, segment);
+};
+
+/**
+ * Response handler for segment-requests being sure to set the correct
+ * property depending on whether the segment is encryped or not
+ * Also records and keeps track of stats that are used for ABR purposes
+ *
+ * @param {Object} segment - a simplified copy of the segmentInfo object from SegmentLoader
+ * @param {Function} finishProcessingFn - a callback to execute to continue processing
+ *                                        this request
+ */
+const handleSegmentResponse = (segment, finishProcessingFn) => (error, request) => {
+  const errorObj = handleErrors(error, request);
+
+  if (errorObj) {
+    return finishProcessingFn(errorObj, segment);
+  }
+  segment.stats = getRequestStats(request);
+
+  if (segment.key) {
+    segment.encryptedBytes = new Uint8Array(request.response);
+  } else {
+    segment.bytes = new Uint8Array(request.response);
+  }
+
+  return finishProcessingFn(null, segment);
+};
+
+/**
+ * Decrypt the segment via the decryption web worker
+ *
+ * @param {WebWorker} decrypter - a WebWorker interface to AES-128 decryption routines
+ * @param {Object} segment - a simplified copy of the segmentInfo object from SegmentLoader
+ * @param {Function} doneFn - a callback that is executed after decryption has completed
+ */
+const decryptSegment = (decrypter, segment, doneFn) => {
+  const decryptionId = 'segment-request-' + Math.random().toString(36);
```

Comment:
Why not use `segment.requestId` here instead of generating another random string?